kclvm/compiler_base/error/src/diagnostic_handler.rs

```diff
@@ -0,0 +1,101 @@
+use crate::{
+    diagnostic::diagnostic_message::TemplateLoader, Diagnostic, DiagnosticStyle, Emitter,
+    MessageArgs, TerminalEmitter,
+};
+use anyhow::{Context, Result};
+use compiler_base_span::fatal_error::FatalError;
+use std::sync::Arc;
+
+pub(crate) struct DiagnosticHandlerInner {
+    emitter: Box<dyn Emitter<DiagnosticStyle>>,
+    diagnostics: Vec<Diagnostic<DiagnosticStyle>>,
```

Comment:
How about `IndexSet` instead of `Vec`

Reply:
`Diagnostic` holds `&dyn`, so `trait hash` can be implemented for `Diagnostic`.
Added comments for duplicate diagnostics.